File: server/app/simulation/race_engine/region.py
"""Port of Region.ts -- half-open interval [start, end) set algebra."""

from __future__ import annotations

from typing import Callable, Union


class Region:
    __slots__ = ("start", "end")

    def __init__(self, start: float, end: float):
        self.start = start
        self.end = end

    def intersect(self, other: "Region") -> "Region":
        start = max(self.start, other.start)
        end = min(self.end, other.end)
        if end <= start:
            return Region(-1, -1)
        return Region(start, end)

    def fully_contains(self, other: "Region") -> bool:
        return self.start <= other.start and self.end >= other.end

    def __repr__(self) -> str:
        return f"Region({self.start!r}, {self.end!r})"

    def __eq__(self, other) -> bool:
        return isinstance(other, Region) and self.start == other.start and self.end == other.end


class RegionList(list):
    def rmap(self, f: Callable[["Region"], Union["Region", list["Region"]]]) -> "RegionList":
        out = RegionList()
        for r in self:
            newr = f(r)
            if isinstance(newr, list):
                for nr in newr:
                    if nr.start > -1:
                        out.append(nr)
            else:
                if newr.start > -1:
                    out.append(newr)
        return out

    def union(self, other: "RegionList") -> "RegionList":
        u = list(self) + list(other)
        r = RegionList()
        if len(u) == 0:
            return r
        u.sort(key=lambda x: x.start)

        acc = u[0]
        for b in u[1:]:
            a = acc
            if a.fully_contains(b):
                acc = a
            elif a.start <= b.start < a.end:
                acc = Region(a.start, b.end)
            elif a.start < b.end <= a.end:
                acc = Region(b.start, a.end)
            else:
                r.append(a)
                acc = b
        r.append(acc)
        return r

File: server/race_engine_rs/src/region.rs
//! Port of region.py -- half-open interval [start, end) set algebra.

#[derive(Clone, Copy, Debug, PartialEq)]
pub struct Region {
    pub start: f64,
    pub end: f64,
}

impl Region {
    pub fn new(start: f64, end: f64) -> Self {
        Region { start, end }
    }

    pub fn intersect(&self, other: &Region) -> Region {
        let start = self.start.max(other.start);
        let end = self.end.min(other.end);
        if end <= start {
            Region::new(-1.0, -1.0)
        } else {
            Region::new(start, end)
        }
    }

    pub fn fully_contains(&self, other: &Region) -> bool {
        self.start <= other.start && self.end >= other.end
    }
}

/// Port of RegionList.rmap/union. Plain `Vec<Region>` elsewhere; these two
/// operations get free functions since Rust has no subclassing to hang them
/// off of.
pub fn rmap_one(regions: &[Region], f: impl Fn(&Region) -> Region) -> Vec<Region> {
    let mut out = Vec::new();
    for r in regions {
        let newr = f(r);
        if newr.start > -1.0 {
            out.push(newr);
        }
    }
    out
}

/// `rmap` when `f` maps one input region to several output regions (e.g.
/// intersecting against a list of corners/slopes/straights).
pub fn rmap_multi(regions: &[Region], f: impl Fn(&Region) -> Vec<Region>) -> Vec<Region> {
    let mut out = Vec::new();
    for r in regions {
        for nr in f(r) {
            if nr.start > -1.0 {
                out.push(nr);
            }
        }
    }
    out
}

pub fn union(a: &[Region], b: &[Region]) -> Vec<Region> {
    let mut u: Vec<Region> = a.iter().chain(b.iter()).copied().collect();
    let mut r = Vec::new();
    if u.is_empty() {
        return r;
    }
    // Python's list.sort is stable; f64 has a total order here since region
    // bounds are never NaN, so partial_cmp().unwrap() matches sorted(key=...).
    u.sort_by(|x, y| x.start.partial_cmp(&y.start).unwrap());

    let mut acc = u[0];
    for &b in &u[1..] {
        let a = acc;
        if a.fully_contains(&b) {
            acc = a;
        } else if a.start <= b.start && b.start < a.end {
            acc = Region::new(a.start, b.end);
        } else if a.start < b.end && b.end <= a.end {
            acc = Region::new(b.start, a.end);
        } else {
            r.push(a);
            acc = b;
        }
    }
    r.push(acc);
    r
}
